Fallback to approvers as reviewers

grantr opened the pull request:
If the newly added blunderbuss option `ApproverFallback` is true, the reviewer selection method will attempt to add approvers to the list if there aren't enough qualifying reviewers to satisfy the requested reviewer count.

While the implementation works and is tested with mixed approver and reviewer lists, this feature is most useful for projects that don't have a reviewer role and thus have OWNERS files containing only approver lists. Those projects should now enable the `ApproverFallback` option to have Blunderbuss select reviewers from approver lists.

/kind feature
/area prow

/assign @cjwagner
/cc @rsdcastro @fejta

Fixes #7691

fejta commented:
Can we just make the review list the union of approvers and reviewers?

kargakis commented:
I would argue for falling back to approvers by default (no config option).

stevekuznetsov commented on prow/plugins/blunderbuss/blunderbuss.go:
```diff
@@ -61,6 +61,9 @@ func helpProvider(config *plugins.Configuration, enabledRepos []string) (*plugin
 }
 
 type ownersClient interface {
```
Seems like you should provide a different implementation of the `ownersClient` that returns approvers instead of duplicating methods

grantr replied:
:+1: I'll experiment with that approach.

grantr replied:
@stevekuznetsov RFAL.

stevekuznetsov commented:
/ok-to-test

fejta commented:
Please get this to pass tests?

grantr commented:
I haven't had time to work on this recently. :( If someone else wants to
pick it up, feel free. Otherwise I'll try to carve out some time next week.

On Sat, May 12, 2018, 11:12 AM Erick Fejta <notifications@github.com> wrote:

> Please get this to pass tests?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/test-infra/pull/7742#issuecomment-388573058>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAACqJVBbD7g2N3GvFYQ0hL-PlnljHIWks5txyYPgaJpZM4TZFiE>
> .
>

grantr commented:
> Can we just make the review list the union of approvers and reviewers?

> I would argue for falling back to approvers by default (no config option).

Where should the fallback live? Should the behavior of `repoowners.RepoOwners` change to always fallback by default? Should blunderbuss wrap `RepoOwners` with its specific union implementation?

fejta commented:
If we make the reviewer set the combination of everyone in reviewers and approvers then we do not need a fallback option, right?

cjwagner commented:
> If we make the reviewer set the combination of everyone in reviewers and approvers then we do not need a fallback option, right?

I don't think we should do that. We want to prioritize using the reviewers over the approvers and only use the approvers as a last resort if no reviewers are found. If we union the approvers and reviewers we can't distinguish them.

I think the fallback belongs in blunderbuss. Not all potential consumers of OWNERS file data from `repoowners` would want to have approvers returned in place of reviewers if reviewers are absent.

grantr commented:
I removed the `GetApprovers` method and replaced it with an implementation that delegates to `Approver` methods. The existing `ownersClient` was renamed to `reviewersClient`:

```go
type reviewersClient interface {
	FindReviewersOwnersForFile(path string) string
	Reviewers(path string) sets.String
	LeafReviewers(path string) sets.String
}
```

`ownersClient` was extended to include those methods plus the Approver methods:

```go
type ownersClient interface {
	reviewersClient
	FindApproverOwnersForFile(path string) string
	Approvers(path string) sets.String
	LeafApprovers(path string) sets.String
}
```

That allows this implementation of `reviewersClient` to be passed to `GetReviewers`:

```go
type fallbackReviewersClient struct {
	ownersClient
}

func (foc fallbackReviewersClient) FindReviewersOwnersForFile(path string) string {
	return foc.ownersClient.FindApproverOwnersForFile(path)
}

func (foc fallbackReviewersClient) Reviewers(path string) sets.String {
	return foc.ownersClient.Approvers(path)
}

func (foc fallbackReviewersClient) LeafReviewers(path string) sets.String {
	return foc.ownersClient.LeafApprovers(path)
}
```

grantr commented:
What's the consensus on whether this should be enabled by default? Seems like @fejta, @kargakis, and @cblecker think it should be the default. @cjwagner WDYT?

cblecker commented:
Yes to this as the default, provided that "no" we don't just union reviewers and approvers. Agree 100% with @cjwagner's [comment](https://github.com/kubernetes/test-infra/pull/7742#issuecomment-390052541). I haven't reviewed the PR/implementation, but just an opinion from the experience POV.

cblecker commented:
Prow as a software product doesn’t need to dictate that, though. That situation would be solved by having that user in both the reviewer *and* the approver section (what we do in most places anyways).

This allows us to have the flexibility to allow a particular overloaded approver to drop from the frontline pool without revoking their approval rights.

grantr commented:
> Yes to this as the default, provided that "no" we don't just union reviewers and approvers.

Currently the code implements the strategy described by @cjwagner in https://github.com/kubernetes/test-infra/pull/7742#issuecomment-390052541. It chooses reviewers first, then approvers only if there aren't enough reviewers.

The original intent of this change is to enable a new workflow that uses approvers without reviewers. I've only considered it carefully in that context. I'm a bit concerned about any unforeseen effects it could have on projects with differing approvers and reviewers lists.

I'm also not sure how best to implement a `true` default. Are there any examples of that? Should the meaning of the option be flipped to `ExcludeApprovers` instead?

cjwagner commented:
> I'm also not sure how best to implement a true default. Are there any examples of that? Should the meaning of the option be flipped to ExcludeApprovers instead?

We usually try to make the default value of the type the default value for the option when possible. It is convenient when the empty value of a config struct is implicitly defaulted properly.
So something like `ExcludeApprovers` or `OnlyReviewers` is preferable IMO.

fejta commented:
> That situation would be solved by having that user in both the reviewer and the approver section

Feels over-optimized for the approvers are not reviewers case, which ought to be rare rather than the default.

We could accomplish the same thing with a field in the `OWNERS` file like a:
```yaml
reviewers:
- him
- her
approvers:
- other

separate-approvers-reviewers-lists: true
```

or alternatively:

```yaml
reviewers:
- him
- her
approvers:
- good-person
only-approver:
- necktie-who-cant-close-vim
```

This exposes the same functionality while a) nudging people in a productive direction and b) requiring less boilerplate for the most common scenarios

grantr commented:
I changed the name of the config option to `ExcludeApprovers` so the default behavior is now to fall back to approvers when necessary. I also cleaned up the duplicate fixtures in the tests so there's only one copy of each test case. Minor unrelated change: a warning is now logged when `maxReviewers` limits the number of reviewers requested, because I found that useful when debugging tests. RFAL.

@fejta I agree with you that making this explicit in OWNERS is probably a better solution, but a much more extensive one.

cblecker commented:
On this PR: it looks good to me. Falling back seems like a sane default. Should probably squash out commits before merge.

On the social friction bikeshed: I agree from a healthy ecosystem point of view, but I'm not convinced that embedding that prow is needed. Maintaining healthy reviewer/approver pools (ensuring reviewers and approvers are active and responsive) is more critical to me.

grantr commented:
Rebased.

fejta commented on prow/plugins/blunderbuss/blunderbuss.go:
```diff
@@ -107,12 +131,28 @@ func handle(ghc githubClient, oc ownersClient, log *logrus.Entry, reviewerCount,
 		}
 		if missing := *reviewerCount - len(reviewers); missing > 0 {
-			log.Warnf("Not enough reviewers found in OWNERS files for files touched by this PR. %d/%d reviewers found.", len(reviewers), reviewerCount)
+			log.Warnf("Not enough reviewers found in OWNERS files for files touched by this PR. %d/%d reviewers found.", len(reviewers), *reviewerCount)
```
should this only warn `if excludeApprovers`?

grantr replied:
I moved this to warn after fallback to approvers (if enabled). WDYT?

fejta commented on prow/plugins/blunderbuss/blunderbuss.go:
```diff
@@ -107,12 +131,28 @@ func handle(ghc githubClient, oc ownersClient, log *logrus.Entry, reviewerCount,
 
 	if len(reviewers) > 0 {
 		if maxReviewers > 0 && len(reviewers) > maxReviewers {
+			log.Warnf("Limiting request of %d reviewers to %d maxReviewers.", len(reviewers), maxReviewers)
```
This seems more like info to me

grantr replied:
Agreed, fixed.

grantr commented:
/retest

grantr commented:
/retest

fejta commented:
/lgtm

fejta commented:
/hold cancel

grantr commented:
Should this be announced on any mailing lists, since it changes default behavior?

fejta commented:
Sent a note to https://groups.google.com/d/msg/kubernetes-sig-contribex/HnOdQ2ChBR4/8QSFCyGKAgAJ
